rebac-admin-backend/v1/response.go

```diff
@@ -25,47 +25,45 @@ func writeErrorResponse(w http.ResponseWriter, r *http.Request, err error) {
 }
 
 // mapErrorResponse returns a Response instance filled with the given error.
-func mapErrorResponse(err error) resources.Response {
+func mapErrorResponse(err error) *resources.Response {
 	if isBadRequestError(err) {
-		return resources.Response{
+		return &resources.Response{
 			Status:  http.StatusBadRequest,
-			Message: "Bad request",
+			Message: err.Error(),
 		}
 	}
 
+	var status int
+
 	switch err.(type) {
 	case *UnauthorizedError:
-		return resources.Response{
-			Status:  http.StatusUnauthorized,
-			Message: "Unauthorized",
-		}
+		status = http.StatusUnauthorized
 	case *NotFoundError:
-		return resources.Response{
-			Status:  http.StatusNotFound,
-			Message: "Not found",
-		}
+		status = http.StatusNotFound
 	default:
-		return resources.Response{
-			Status:  http.StatusInternalServerError,
-			Message: "Unexpected error",
-		}
+		status = http.StatusInternalServerError
+	}
+
+	return &resources.Response{
+		Message: err.Error(),
+		Status:  status,
 	}
 }
 
-// isBadRequestError determines whether the given error should be teated as a
-// "Bad Request" (400) error.
-func isBadRequestError(err error) bool {
-	switch err.(type) {
-	case *resources.UnmarshalingParamError:
-		return true
-	case *resources.RequiredParamError:
-		return true
-	case *resources.RequiredHeaderError:
-		return true
-	case *resources.InvalidParamFormatError:
-		return true
-	case *resources.TooManyValuesForParamError:
-		return true
+// writeResponse is a helper method to avoid verbose repetition of very common instructions
+// it writes to the ResponseWriter either:
+// - the provided marshalledResponse if marshalErr is nil
+// - or the mapped error response based on marshalErr if it is not nil
+func writeResponse(w http.ResponseWriter, status int, marshalledResp []byte, marshalErr error) {
```

Comment:
suggestion:
```
func writeResponse(w http.ResponseWriter, status int, responseObject interface{}) {
        data, err := json.Marshal(responseObject)
        if err != nil {
		writeErrorResponse(w, err)
		return
	}

	w.WriteHeader(status)
	_, err := w.Write(data)
	if err != nil {
		writeErrorResponse(w, err)
	}
}
```
which is a bit more general in that it removes the json marshalling step from all the functions that use it

Reply:
Thanks Ales, I was actually undecisive about this and waited for review time to gather thoughts
I'll go that way, next revision coming soon